azure-kusto-data/source/tokenProvider.ts

```diff
@@ -0,0 +1,313 @@
+            const suffix = this.kustoUri.endsWith("/") ? ".default" : "/.default";
+            this.scopes = [kustoUri + suffix];
+        }
+    }
+}
+
+/**
+ * Basic Token Provider keeps and returns a token received on construction
+ */
+export class BasicTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    token: string;
+
+    constructor(kustoUri: string, token: string) {
+        super(kustoUri);
+        this.token = token;
+    }
+
+    acquireToken(): Promise<TokenResponse> {
+        return Promise.resolve<TokenResponse>({ tokenType: BEARER_TYPE, accessToken: this.token });
+    }
+}
+
+/**
+ * Callback Token Provider generates a token based on a callback function provided by the caller
+ */
+export class CallbackTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    callback: () => Promise<string>;
+
+    constructor(kustoUri: string, callback: () => Promise<string>) {
+        super(kustoUri);
+        this.callback = callback;
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        const token = await this.callback();
+        return { tokenType: BEARER_TYPE, accessToken: token };
+    }
+}
+
+
+// MSI Token Provider obtains a token from the MSI endpoint
+// The args parameter is a dictionary conforming with the ManagedIdentityCredential initializer API arguments
+export class MsiTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    clientId: string;
+    managedIdentityCredential!: ManagedIdentityCredential;
+
+    constructor(kustoUri: string, clientId: string) {
+        super(kustoUri);
+        this.clientId = clientId;
+    }
+
+    init(): void {
+        return;
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        if (this.managedIdentityCredential == null) {
+            this.managedIdentityCredential = new ManagedIdentityCredential(this.clientId);
+        }
+        const msiToken = await this.managedIdentityCredential.getToken(this.kustoUri);
+        if (msiToken?.token != null) {
+            return { tokenType: BEARER_TYPE, accessToken: msiToken.token };
+        }
+        throw new Error(`"Failed to obtain MSI token for '${this.kustoUri}' with '${this.clientId}'`)
+    }
+}
+
+/**
+ * AzCli Token Provider obtains a refresh token from the AzCli cache and uses it to authenticate with MSAL
+ */
+export class AzCliTokenProvider extends TokenProviderBase {
+    context = { "authority": this.name };
+    name = "CallbackTokenProvider";
+    azureCliCredentials!: AzureCliCredentials;
+
+    constructor(kustoUri: string) {
+        super(kustoUri);
+    }
+
+    async acquireToken(): Promise<TokenResponse> {
+        if (this.azureCliCredentials == null) {
+            this.azureCliCredentials = await AzureCliCredentials.create({ resource: this.kustoUri });
```

Comment:
Do you want to introduce a call to CloudSettings here in case this is used to call Synapse?

Reply:
Talked with Ran, he said I should wait with it (currently also the C# client doesn't support this scenario)

Reply:
Seems like a simple change.
Why not support it?

Reply:
According to Ran it's not... we can sync about it with him

Reply:
Let's.
IIRC all we need to do is replace the URI with the target resource URI from CloudSettings